File: Java/src/FindWord.java
import java.util.*;

public class FindWord {
    public static void main(String[] args) {

        // Test cases
        List<String> test1 = Arrays.asList("P>E", "E>R", "R>U");
        System.out.println(findWord(test1)); // Should print "PERU"

        List<String> test2 = Arrays.asList("I>N", "A>I", "P>A", "S>P");
        System.out.println(findWord(test2)); // Should print "SPAIN"

        List<String> test3 = Arrays.asList("U>N", "G>A", "R>Y", "H>U", "N>G", "A>R");
        System.out.println(findWord(test3)); // Should print "HUNGARY"

        List<String> test4 = Arrays.asList("I>F", "W>I", "S>W", "F>T");
        System.out.println(findWord(test4)); // Should print "SWIFT"

        List<String> test5 = Arrays.asList("R>T", "A>L", "P>O", "O>R", "G>A", "T>U", "U>G");
        System.out.println(findWord(test5)); // Should print "PORTUGAL"

        List<String> test6 = Arrays.asList("W>I", "R>L", "T>Z", "Z>E", "S>W", "E>R", "L>A", "A>N", "N>D", "I>T");
        System.out.println(findWord(test6)); // Should print "SWITZERLAND"

        List<String> test7 = List.of("O>K");
        System.out.println(findWord(test7)); // Should print "OK"

    }

    public static String findWord(List<String> combinations) {
        // Create a map to store the edges (currentChar -> nextChar)
        Map<Character, Character> edgeMap = new HashMap<>();
        // Create a set to keep track of all destination characters
        Set<Character> destinations = new HashSet<>();

        // Populate the edge map and destination set
        for (String combo : combinations) {
            char from = combo.charAt(0);
            char to = combo.charAt(2);
            edgeMap.put(from, to);
            destinations.add(to);
        }

        // Find the starting character (the one that isn't a destination)
        char start = ' ';
        for (String combo : combinations) {
            char candidate = combo.charAt(0);
            if (!destinations.contains(candidate)) {
                start = candidate;
                break;
            }
        }

        // Reconstruct the word by following the edges
        StringBuilder word = new StringBuilder();
        char current = start;
        while (current != '\0') {
            word.append(current);
            current = edgeMap.getOrDefault(current, '\0');
        }

        return word.toString();
    }

}

File: Java/kotlin/FindWord.kt
package kotlin

object FindWord {
    @JvmStatic
    fun main(args: Array<String>) {
        // Test cases
        val test1 = listOf("P>E", "E>R", "R>U")
        println(findWord(test1)) // Should print "PERU"

        val test2 = listOf("I>N", "A>I", "P>A", "S>P")
        println(findWord(test2)) // Should print "SPAIN"

        val test3 = listOf("U>N", "G>A", "R>Y", "H>U", "N>G", "A>R")
        println(findWord(test3)) // Should print "HUNGARY"

        val test4 = listOf("I>F", "W>I", "S>W", "F>T")
        println(findWord(test4)) // Should print "SWIFT"

        val test5 = listOf("R>T", "A>L", "P>O", "O>R", "G>A", "T>U", "U>G")
        println(findWord(test5)) // Should print "PORTUGAL"

        val test6 = listOf("W>I", "R>L", "T>Z", "Z>E", "S>W", "E>R", "L>A", "A>N", "N>D", "I>T")
        println(findWord(test6)) // Should print "SWITZERLAND"

        val test7 = listOf("O>K")
        println(findWord(test7)) // Should print "OK"
    }

    fun findWord(combinations: List<String>): String {
        // Create a map to store the edges (currentChar -> nextChar)
        val edgeMap = mutableMapOf<Char, Char>()
        // Create a set to keep track of all destination characters
        val destinations = mutableSetOf<Char>()

        // Populate the edge map and destination set
        for (combo in combinations) {
            val from = combo[0]
            val to = combo[2]
            edgeMap[from] = to
            destinations.add(to)
        }

        // Find the starting character (the one that isn't a destination)
        var start = ' '
        for (combo in combinations) {
            val candidate = combo[0]
            if (candidate !in destinations) {
                start = candidate
                break
            }
        }

        // Reconstruct the word by following the edges
        val word = StringBuilder()
        var current: Char? = start
        while (current != null && current != '\u0000') {
            word.append(current)
            current = edgeMap[current]
        }

        return word.toString()
    }
}

